clients/callback_client.go

```diff
@@ -27,113 +33,173 @@ func NewCallbackClient() CallbackClient {
 		Timeout: 5 * time.Second, // Give up on requests that take more than this long
 	}
 
-	return CallbackClient{
-		httpClient: client.StandardClient(),
+	return &PeriodicCallbackClient{
+		httpClient:               client.StandardClient(),
+		callbackInterval:         callbackInterval,
+		requestIDToLatestMessage: map[string]TranscodeStatusMessage{},
+		mapLock:                  sync.RWMutex{},
 	}
 }
 
-func (c CallbackClient) DoWithRetries(r *http.Request) error {
-	// TODO: Replace with a proper shared Secret, probably coming from the initial request
-	r.Header.Set("Authorization", "Bearer IAmAuthorized")
-
-	resp, err := c.httpClient.Do(r)
-	if err != nil {
-		return fmt.Errorf("failed to send callback to %q. Error: %s", r.URL.String(), err)
-	}
-	defer resp.Body.Close()
-
-	if resp.StatusCode >= 400 {
-		return fmt.Errorf("failed to send callback to %q. HTTP Code: %d", r.URL.String(), resp.StatusCode)
-	}
-
-	return nil
+// Start looping through all active jobs, sending a callback for the latest status of each
+// and then pausing for a set amount of time
+func (pcc *PeriodicCallbackClient) Start() *PeriodicCallbackClient {
+	go func() {
+		for {
+			recoverer(pcc.SendCallbacks)
+			time.Sleep(pcc.callbackInterval)
+		}
+	}()
+	return pcc
 }
 
-func (c CallbackClient) SendRecordingEvent(event *RecordingEvent) {
-	eventJson, err := json.Marshal(event)
-	if err != nil {
-		log.Printf("SendRecordingStarted json marshal %v", err)
-		return
-	}
-	req, err := http.NewRequest(http.MethodPost, config.RecordingCallback, bytes.NewReader(eventJson))
-	if err != nil {
-		log.Printf("SendRecordingStarted http.NewRequest %v", err)
-		return
-	}
-	if err := c.DoWithRetries(req); err != nil {
-		log.Printf("SendRecordingStarted callback %v", err)
-	}
+func recoverer(f func()) {
+	defer func() {
+		if err := recover(); err != nil {
+			log.LogNoRequestID("panic in callback goroutine, recovering", "err", err)
+			go recoverer(f)
```

Comment:
Won't this create an eternal panic loop calling `f` in background? 🤔

Reply:
Let's continue the discussion under https://github.com/livepeer/catalyst-api/pull/219